### DISCUSSION TITLE
What is the behaviour of the multivariate forecasting in the following cases and why?

### DISCUSSION THREAD

**Discussion by parissenseone** [alan-turing-institute/sktime#5572] [Q&A]
**The case is I have a y lets say y is energy, i have also t which is lets say temperature and i have also some static parameters that can apply per row which are not temporal so i dont want windows on them and i dont want to repeat them in the windows.**

**In that case previous values of y(energy) will be used in windows as expected. The same can apply to t(temperature). Also i can use date features from y (i believe this can be done now , how??) Now, i want to apply on this non temporal vectors per row. So we have a vector in each row with [window(y[-1...-12]), window(t[-1...-12]), non temporal feature vector, date features, known forecasted features(exogenus)] and then predict y(energy for steps 0..6) knowing the date features of the future(easily) and some forecasted features(exogenus).**

**How someone will approach this in sktime without using forecasterX or forecasting variables to use them as forecasters??? 
I want to do exactly this**

### Scenarios I have tested

**I want to have the following scenarios:**

1. a and b with windows of 12 to predict a for next 6 (so 12 a,b (1 to 12) and predict 3 a (13,18))
2. a and b with windows of 12 to predict a for next 6 but knowing b (so 12 a,b (1 to 12) and predict 3 a (13,18) knowing b(13,18))
    
I have made the following experiments:

I will run both of them in multiouput and recursive strategy :    
```from xgboost import XGBRegressor
from sktime.datasets import load_airline
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.naive import NaiveForecaster
from sktime.forecasting.compose import make_reduction, TransformedTargetForecaster
from random import random
import pandas as pd
import numpy as np

y = load_airline()
X = pd.DataFrame({'month': y.index.month}, index=y.index)
E = pd.DataFrame({'month': y.index.month}, index=y.index)
forecaster = make_reduction(
    XGBRegressor(objective='reg:squarederror', random_state=42, n_jobs=3,nthread=3,tree_method = "hist"), 
    window_length=12, strategy='multioutput'
)
fh = ForecastingHorizon(np.arange(1, 7), is_relative=True)
```

Now the expriments:

No X at all
```
forecaster.fit(y=y, fh=fh)
y_pred0 = forecaster.predict(X=None)
```

Nan X in fit
```
E = pd.DataFrame({'month': y.index.month}, index=y.index)
E['month'] = np.nan
forecaster.fit(y=y, X=E, fh=fh)
y_pred1 = forecaster.predict(X=E)
```

X in fit but NaN X in predict
```
E = pd.DataFrame({'month': y.index.month}, index=y.index)
E['month'] = np.nan
forecaster.fit(y=y, X=X, fh=fh)
y_pred2 = forecaster.predict(X=E)
```

X in fit and predict giving already known values of X in the future
```
E = pd.DataFrame({'month': y.index.month}, index=y.index)
forecaster.fit(y=y, X=E, fh=fh)
new_Erows = pd.DataFrame({
    'month': [4,3,2,1,5,1]}, 
    index=[
        pd.Period('1961-01', freq='M'),
        pd.Period('1961-02', freq='M'),
        pd.Period('1961-03', freq='M'),
        pd.Period('1961-04', freq='M'),
        pd.Period('1961-05', freq='M'),
        pd.Period('1961-06', freq='M')
    ])
E = pd.concat([E, new_Erows])
y_pred3 = forecaster.predict(X=E[-6:])
```

X in fit and predict giving already known values of X in the future but all 0 except the latest
```
E = pd.DataFrame({'month': y.index.month}, index=y.index)
E['month'] = np.nan
forecaster.fit(y=y, X=X, fh=fh)
new_Erows = pd.DataFrame({
    'month': [0,0,0,0,0,12]}, 
    index=[
        pd.Period('1961-01', freq='M'),
        pd.Period('1961-02', freq='M'),
        pd.Period('1961-03', freq='M'),
        pd.Period('1961-04', freq='M'),
        pd.Period('1961-05', freq='M'),
        pd.Period('1961-06', freq='M')
    ])
E = pd.concat([E, new_Erows])

y_pred4 = forecaster.predict(X=E[-6:])
```

Results for recursive
```
multiou = pd.concat({
    '(M)Only y':y_pred0,
    '(M)NaN X': y_pred1,
    '(M)X but NaN X in predict':y_pred2,
    '(M)X and X-predict with values':y_pred3,
    '(M)X and X-predict with values all 0 except final 12':y_pred4
    },axis=1)
```

Same code applies to multioutput.

The results are:

<img width="1360" alt="image" src="https://github.com/sktime/sktime/assets/127109390/3fe1426c-09e9-466f-83a7-0992690dbb24">

We see:
**Expected:**
1. "`(R)Only y`" | "`(R)NaN X`" are the same as expected.
2. "`(M)Only y`" | "`(M)NaN X`" are the same as expected.
3. "`(R)X but NaN X in predict`" is different from "`(R)Only y and (R)NaN`" as expected

**Need answer**
1. "`(R)X and X-predict with values`" is different from "`(R)X but NaN X`" (expected but why not happening in multioutput)
2. "`(R)X and X-predict with values all 0 except final 12`" is exactly the same as "`(R)X and X-predict with values`" (expected??)

**Weird Need answer**
1. "`(M)X but NaN X in predict`" , "`(M)X and X-predict with values`" and  "`(M)X and X-predict with values all 0 except final 12`" are exactly the same! Why????? The exogenous variable is not working on multioutput?

Is there any issue with the code? or i miss something?

**Comment by fkiraly** [comment]
#### reduction forecaster behaviour

FYI @Benheid, if I remember correctly, you have looked into this most recently, in detail.


#### Forecasting `y` without forecasting `X`

> How someone will approach this in sktime without using forecasterX or forecasting variables to use them as forecasters

I think to answer this question, we first need to make it precise - I would like ot understand better what you are looking for.

If I were to formally phrase your question, it would be "how to produce `y` forecasts if you do not have future `X`" - but you do not seem to like one particular example, so it feels a condition for the desired solution is missing.

On "how to produce `y` forecasts if you do not have future `X`":

`sktime` is a library in the sense that new forecasters can be added easily in a compatible interface, and they can be combined with each others to build other forecasters. Pipelines and composites also behave as forecasters.

Currently there are 70 forecasters directly in `sktime`, and you can build a theoretical infinity out of compositors and transformer components (realistically, the number of variants that might make sense here is a very large but not infinite number).

So, naively I would read your question as the more precise one: "which forecasters out of those that you can build in principle do not require `X` known in the future".

The typical example is that which @yarnabrina has given on discord:

* most forecasters *do require `X` to be known in the future*. This is not a limitation of `sktime` as a framework, but a limitation of the forecasters which often also come from other packages.
* the most direct way to turn these forcasters into ones that do not require `X` in the future is to use `ForecastX`, to forecast unknown variables, because then you can apply any forecasters that needs that `X`.
* using `ForecastX` produces a new forecaster, which does not require future `X` - it produces a specification, which is a forecaster of the desired type that you built out of multiple components, not data primarily.

If your question would be: is there a forecaster in `sktime` that does not require future `X` to make an `y` forecast - I would not be aware of such a forecaster, but I may be mistaken. Unfortunately afaik there is not a tag which makes this a searchable property, perhaps that is an action out of this?

FYI @benHeid, @yarnabrina, @marrov, @arnaujc91, in case you have sth to add.

**Comment by parissenseone** [comment]
Yes the concept is that: (and of course someone can write his own models based on the base forecaster of sktime but i am talking about the existing models behaviour)

**How someone will approach multivariate input and one output in Sktime without using forecasterX or forecasting variables to use them as forecasters?? With XGboost for example as the above experiments**. 

### There is a strange behaviour in a specific case in the above experiments.

**"`(M)X but NaN X in predict`" , "`(M)X and X-predict with values`" and "`(M)X and X-predict with values all 0 except final 12`" are exactly the same!** 
**Why??** 
**The exogenous variable is not working on multioutput? Meaning, exogenous variables are not used in predict in such case?**

- It seems that the X is working for fit in that cases but not used in predict at all. Is this something normal? 
- So, there are models implemented in sktime that you dont use the forecasting features if they exist and others that you use them?

- In this specific case it seems that indeed we have x,y for train and produce y and x is not used in predict at all even though it is needed as a function parameter. (the hack of nans).

**!! A question also lies here: Ok the y is windowed, what is happening to x in those cases? Are they windowed? (in the specific implementation, not if we can windowed them). And if not how we can windowed also x?**

**Reply by parissenseone** [reply]
Is there any news on that?
